### PR TITLE
tools: Downstream tap scripts changes

### PR DESCRIPTION
**Pull request by jtojnar**
These are changes from libsecret and gnome-keyring, specifically the following commits:

* https://gitlab.gnome.org/GNOME/gnome-keyring/commit/918ab836fd23f9b2c0bf655c7f4f575bffff3189
* https://gitlab.gnome.org/GNOME/gnome-keyring/commit/7e75a62e846551c19b9c875f7b237a07ee3b93e1
* https://gitlab.gnome.org/GNOME/gnome-keyring/commit/7120f44ceedd5c10802781d64a5829b3c6d8e13f
* https://gitlab.gnome.org/GNOME/libsecret/commit/98d12ea2201ec5b52e561e06b3af85be95fcb482

cc @ueno

### CODE REVIEW

File: tools/tap-driver [hunk 1]
```diff
@@ -284,9 +284,12 @@ class TapDriver(Driver):
             if code != "SKIP":
                 skipped = False
 
-        if not self.errored and returncode:
-            self.report_error("process failed: %d" % returncode)
-            self.errored = True
+        if not self.errored:
+            if returncode == 77:
+                skipped = True
```
Comment (martinpitt) [inline tools/tap-driver]: So exit code 77 is a magic one that means "skip"? I don't see that in the [TAP specification](http://testanything.org/tap-specification.html), and a [Google search](https://www.google.com/search?q=tap+exit+code+77) isn't that enlightening either. Can you please add a reference/comment for that?

File: tools/tap-gtester [hunk 2]
```diff
@@ -90,6 +90,8 @@ class GTestCompiler(NullCompiler):
                        print("ok %d %s" % (self.test_num, self.test_name))
                    if data == "FAIL":
                        print("not ok %d %s" % (self.test_num, self.test_name))
+                   if data == "SKIP":
+                       print("ok %d %s # skip" % (self.test_num, self.test_name))
```
Comment (martinpitt) [inline tools/tap-gtester]: The code change makes sense. Just the commit message doesn't really fit for Cockpit, there is no test-pam.c .

File: tools/tap-driver [hunk 3]
```diff
@@ -212,8 +212,11 @@ class TapDriver(Driver):
             return
         description = description.lstrip()
 
-        # Special case if description starts with this, then skip
-        if description.lower().startswith("# skip"):
+        # Parse out a directive from description, if any
+        (description, unused, directive) = description.partition("#")
+
+        # Special case if directive starts with this, then skip
+        if directive.lstrip().lower().startswith("skip"):
```
Comment (martinpitt) [inline tools/tap-driver]: This looks wrong. This function starts out with a whole line like

    ok 23 # skip known to not work here

which the original code turned into `description == "# skip known to not work here"`, and pass that on to `result_skip()`. But the new `partition()` call above now turns this into `description == ""` (which is passed on), and `directive == "skip known to not work here"` (which is thrown away).

I'm not sure what this is actually supposed to fix. Can you please be more verbose in the commit log? Perhaps you want to split this as

    (_, _, description) = description.partition("#")

? The visible effect of that would be to drop the `#` from the reported result.

File: tools/tap-driver [hunk 4]
```diff
@@ -132,7 +133,10 @@ class Driver:
             self.report("SKIP", *args)
 
     def report_error(self, description=""):
-        self.report("ERROR", "", description)
+        if self.enable_hard_errors:
+            self.report("ERROR", "", description)
+        else:
+            self.result_fail(description)
```
Comment (martinpitt) [inline tools/tap-driver]: The default of that option is `False`, so that changes behaviour. What's the visible effect of that?

File: tools/tap-driver (hunk 3 above)
Comment (ueno) [reply tools/tap-driver]: > which the original code turned into description == "# skip known to not work here"

The problem is that the original code doesn't work like that. After [line 207](https://github.com/cockpit-project/cockpit/pull/9924/files/17e5ead8defe8c145f381acae051b68901e2f68a..821fa7707ac4e4c0029d2d7d870115a7e1530c03#diff-b5a45ed5cac3c210f90850e7d32e0116L207), description becomes: `23 # skip known to not work here`. This second `partition()` is to scrub "23 # ".

File: tools/tap-driver (hunk 4 above)
Comment (ueno) [reply tools/tap-driver]: This enables the use of [XFAIL_TESTS](https://gitlab.gnome.org/GNOME/libsecret/blob/master/.gitlab-ci.yml#L36) when the test suite runs under automake's harness.

File: tools/tap-driver (hunk 1 above)
Comment (stefwalter) [reply tools/tap-driver]: I chose 77 as a way for simple executable tests to communicate the fact that they should be skipped. I still see usage of it in src/retest/retest.c ... although it's used a lot less than before.

File: tools/tap-driver (hunk 3 above)
Comment (martinpitt) [reply tools/tap-driver]: Indeed this is really hard to follow - this should *really* use a regular expression, that's so much easier to write and read, and more robust. Anyway, this *is* broken. Run `make check TESTS=test-locale` on master, you will see
```
SKIP: test-locale 7 # skip -- locale not available
PASS: test-locale 8 /locale/set-language/de-de
invalid or unusable locale: __xx;%%%.UTF-8
SKIP: test-locale 9 # skip -- locale not available
```
with this PR, the description gets killed, as I wrote above:
```
SKIP: test-locale 7 
PASS: test-locale 8 /locale/set-language/de-de
invalid or unusable locale: __xx;%%%.UTF-8
SKIP: test-locale 9 
```

**Comment by martinpitt** [comment]
@jtojnar , @ueno , do you still want to go ahead with this? This causes regressions, so it can't be merged as-is.

**Comment by ueno** [comment]
I would leave it to @jtojnar; in libsecret I rather incline to use the stock tap scripts from automake, until we eventually switch to meson.

**Comment by jtojnar** [comment]
I think I will rather focus on porting projects to meson.
